xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,368 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/resolver"
+)
+
+var (
+	errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+)
+
+// NewBalancer returns a graceful switch Balancer.
+func NewBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) *Balancer {
+	return &Balancer{
+		cc:    cc,
+		bOpts: opts,
+	}
+}
+
+// Balancer is a utility to gracefully switch from one balancer to
+// a new balancer.
+type Balancer struct {
```

Comment:
Could we have a comment on the recommended way to use this utility?

It does not come with a registered `balancer.Builder`, so maybe it is more suited for certain use cases than others? Could we have a comment on which use cases are most appropriate for this utility.

And probably a comment on what interfaces this utility implements. And if this is expected to implement the `balancer.Balancer` interface, does a compile time check for the same make sense?

Reply:
Well, previously, we had envisioned a use case for a Cluster Specifier Plugin which changed type and also a balancer that receives a config change that is different enough that it needs time to warm up the new policy, but wants to continue to serve RPC's. Now, if we use it as the top level LB policy for the channel, this will implicitly capture all use cases. Should I state that it should be used as the top level balancer for the channel? Added a comment saying It implements the balancer.Balancer interface. I don't get what "And if this is expected to implement the balancer.Balancer interface, does a compile time check for the same make sense?" is referring to, could you please elaborate further?

Reply:
Okay, I added one. I kinda get why I added it, but not really. Is this so if balancer API changes over time we will know to change this for maintainability reasons?